### DISCUSSION TITLE
Github Branching Model

### DISCUSSION THREAD

**Discussion by Milind220** [Ozon3Org/Ozon3#24] [General]
This is a great Github Branching model for this repository to follow.

https://nvie.com/posts/a-successful-git-branching-model/

**Comment by Milind220** [comment]
### Summary of branching model

`main` branch
This branch is always stable.
-> No direct commits to this branch. The only commits that are made here are from merging in a `release` or `hotfix` branch.
-> Any commits on this branch are _by definition_ a new release to be published on PyPi. 
-> Infinite lifetime.

`dev` branch
This branch is where most new features are added.
-> Direct commits only from admins (to keep it clean for open-source contributions). PR's welcome from anyone.
-> infinite lifetime.

'feature' branch
A special branch created for a big new feature.
-> branched off from 'dev', and merged back into `dev` when the feature is complete
-> lives as long as feature is being developed

`release` branch
A special branch created from `dev` in preparation for a new release. 
-> This is where final testing and bugfixes take place before being merged into `main` and also `dev`.
-> lives as long as release is being fine tuned.

`hotfix` branch
A special branch to fix bugs that are discovered in released code. 
-> branched from `main`, bugs are fixed, then merged back into `main` as a new release, and into `dev` as well.

**Comment by Milind220** [comment]
After using this for a bit now, I'm experimenting with the following minor edit to this model:

* for a new release, merge release branch into main
* instead of now merging release also into dev, which causes dev to fall one commit behind main, we could merge main into dev, and then dev and main have the same number of commits. This is essentially the same thing as merging release into dev.

You can take a look at the Stack Overflow question that prompted me to make this change [here](https://stackoverflow.com/questions/71435752/when-i-merge-my-release-branch-into-main-and-then-into-dev-to-incorporate-the-c)

**Comment by Samxx97** [comment]
I checked the stack overflow question and this definitely makes more sense , kudos!
